components/script/dom/gpucommandencoder.rs

```diff
@@ -68,18 +92,52 @@ impl GPUCommandEncoderMethods for GPUCommandEncoder {
         &self,
         _descriptor: &GPUComputePassDescriptor,
     ) -> DomRoot<GPUComputePassEncoder> {
-        GPUComputePassEncoder::new(&self.global(), self.channel.clone(), self.encoder)
+        *self.state.borrow_mut() = GPUCommandEncoderState::EncodingComputePass;
+        GPUComputePassEncoder::new(&self.global(), self.channel.clone(), &self)
     }
 
     /// https://gpuweb.github.io/gpuweb/#dom-gpucommandencoder-copybuffertobuffer
     fn CopyBufferToBuffer(
         &self,
         source: &GPUBuffer,
-        source_offset: u64,
+        source_offset: GPUSize64,
         destination: &GPUBuffer,
-        destination_offset: u64,
-        size: u64,
+        destination_offset: GPUSize64,
+        size: GPUSize64,
     ) {
+        let mut valid = match source_offset.checked_add(size) {
+            Some(_) => true,
+            None => false,
+        };
+        valid &= match destination_offset.checked_add(size) {
+            Some(_) => true,
+            None => false,
+        };
+        valid &= match BufferUsage::from_bits(source.usage()) {
+            Some(usage) => usage.contains(BufferUsage::COPY_SRC),
+            None => false,
+        };
+        valid &= match BufferUsage::from_bits(destination.usage()) {
+            Some(usage) => usage.contains(BufferUsage::COPY_DST),
+            None => false,
+        };
+        valid &= (*self.state.borrow() == GPUCommandEncoderState::Open) &&
+            source.valid() &&
+            destination.valid() &
+                !(size & 3 == 0) &
+                !(source_offset & 3 == 0) &
+                !(destination_offset & 3 == 0) &
+                (source.size() >= source_offset + size) &
+                (destination.size() >= destination_offset + size);
+        if source.id().0 == destination.id().0 {
+            valid &= source_offset > destination_offset + size ||
+                source_offset + size < destination_offset;
+        }
+
+        if !valid {
+            return;
```

Comment:
we'd want to mark the whole command encoder as invalid if something bad happens

Reply:
The spec doesn't mention it yet but do we need to record a validation error here? If so I'll add a  TODO.

Reply:
Yes, we definitely do need to record an error in the current error scope here.
But that relies on the error scopes being exposed and implemented, so you don't have to do it all in this PR.